File: 3152-special-array-ii/3152-special-array-ii.py
class Solution:
  def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
    z = 0
    check = [0] * len(nums)

    for i in range(1, len(nums)):
      z += 1
      if nums[i - 1] % 2 == nums[i] % 2:
        z = 0
      check[i] = z

    ans = []
    for f, t in queries:
      if check[t] >= t - f:
        ans.append(True)
      else:
        ans.append(False)
        
    return ans      

File: 3152-special-array-ii/3152-special-array-ii.ts
function isArraySpecial(nums: number[], queries: number[][]): boolean[] {
  const index = new Array(nums.length).fill(0);
  let l = 0;
  for (let r = 1; r < nums.length; r++) {
    if (nums[r] % 2 === nums[r - 1] % 2) l = r;
    index[r] = l;
  }
  const ans = new Array(queries.length).fill(false);
  for (let i = 0; i < queries.length; i++) {
    if (index[queries[i][1]] <= queries[i][0]) ans[i] = true;
  }
  return ans;
};